File: typescript/runkite-runner/src/tenantCtx.ts
/**
 * Per-run tenant + run-binding for direct-mode store SQL and proxy headers.
 * Mirrors python/runkite_runner/tenant_ctx.py -- see that module's docstring.
 *
 * Also owns the LangGraph checkpointer thread-id encoding (see
 * checkpointThreadId) so PostgresSaver rows do not collide across tenants.
 */
import { AsyncLocalStorage } from "node:async_hooks";

const DEFAULT_TENANT = "default";

export type RunBinding = {
  tenantId: string;
  runId: string;
  generation: number;
};

const als = new AsyncLocalStorage<RunBinding>();

/** Header runners send on /internal/* (matches internal/auth.HeaderTenantID). */
export const HEADER_TENANT_ID = "X-Runkite-Tenant-Id";
export const HEADER_RUN_ID = "X-Runkite-Run-Id";
export const HEADER_GENERATION = "X-Runkite-Generation";

export function currentTenant(): string {
  return als.getStore()?.tenantId || DEFAULT_TENANT;
}

export function runWithTenant<T>(tenantId: string | undefined, fn: () => Promise<T>): Promise<T> {
  const tid = (tenantId ?? "").trim() || DEFAULT_TENANT;
  const prev = als.getStore();
  return als.run(
    {
      tenantId: tid,
      runId: prev?.runId ?? "",
      generation: prev?.generation ?? 0,
    },
    fn,
  );
}

/** Bind tenant + run id/generation for the duration of fn (proxy headers). */
export function runWithBinding<T>(
  opts: { tenantId?: string; runId: string; generation?: number },
  fn: () => Promise<T>,
): Promise<T> {
  const tid = (opts.tenantId ?? "").trim() || DEFAULT_TENANT;
  return als.run(
    {
      tenantId: tid,
      runId: (opts.runId ?? "").trim(),
      generation: opts.generation ?? 0,
    },
    fn,
  );
}

export function tenantHeaders(): Record<string, string> {
  const store = als.getStore();
  const headers: Record<string, string> = {
    [HEADER_TENANT_ID]: store?.tenantId || DEFAULT_TENANT,
  };
  if (store?.runId) {
    headers[HEADER_RUN_ID] = store.runId;
    headers[HEADER_GENERATION] = String(store.generation ?? 0);
  }
  return headers;
}

/** LangGraph checkpointer key for this assignment's logical thread.
 * Empty / "default" → bare threadId (existing single-tenant rows stay
 * reachable). Any other tenant → `${tenantId}:${threadId}`. */
export function checkpointThreadId(tenantId: string | undefined | null, threadId: string): string {
  const tid = (tenantId ?? "").trim() || DEFAULT_TENANT;
  if (tid === DEFAULT_TENANT) return threadId;
  return `${tid}:${threadId}`;
}

File: python/runkite_runner/tenant_ctx.py
"""Per-run tenant + run-binding for direct-mode store/vector SQL and proxy headers.

The runner process holds one shared store (and optional vector store)
across concurrent jobs. Multi-tenant isolation requires the active
tenant to follow the job's RunAssignment.tenant_id for the duration of
that job -- a process-wide constant of "default" is wrong. ContextVar
is task-local across await points, so concurrent jobs do not clobber
each other.

Proxy-mode /internal/store|vectors|connectors/{session,mcp} also require
X-Runkite-Run-Id + X-Runkite-Generation so the control plane can derive
tenant/agent from the in-flight assignment instead of trusting the
tenant header alone.

Also owns the LangGraph checkpointer thread-id encoding: non-default
tenants prefix configurable.thread_id so AsyncPostgresSaver rows do not
collide when two tenants reuse the same client-chosen thread id.
"""

from __future__ import annotations

from contextvars import ContextVar, Token
from dataclasses import dataclass

_DEFAULT = "default"
_active: ContextVar[str] = ContextVar("runkite_tenant", default=_DEFAULT)


@dataclass(frozen=True)
class _RunBinding:
    run_id: str
    generation: int


_run: ContextVar[_RunBinding | None] = ContextVar("runkite_run", default=None)

# Header runners send on /internal/* so proxy-mode handlers scope the
# same way as client auth (see internal/auth.HeaderTenantID).
HEADER_TENANT_ID = "X-Runkite-Tenant-Id"
HEADER_RUN_ID = "X-Runkite-Run-Id"
HEADER_GENERATION = "X-Runkite-Generation"


def current_tenant() -> str:
    tid = _active.get()
    return tid if tid else _DEFAULT


def bind_tenant(tenant_id: str | None) -> Token[str]:
    tid = (tenant_id or "").strip() or _DEFAULT
    return _active.set(tid)


def reset_tenant(token: Token[str]) -> None:
    _active.reset(token)


def bind_run(run_id: str | None, generation: int | None) -> Token[_RunBinding | None]:
    rid = (run_id or "").strip()
    if not rid:
        return _run.set(None)
    gen = int(generation or 0)
    return _run.set(_RunBinding(run_id=rid, generation=gen))


def reset_run(token: Token[_RunBinding | None]) -> None:
    _run.reset(token)


def tenant_headers() -> dict[str, str]:
    """Extra headers for proxy-mode HTTP calls to the control plane."""
    headers = {HEADER_TENANT_ID: current_tenant()}
    binding = _run.get()
    if binding is not None:
        headers[HEADER_RUN_ID] = binding.run_id
        headers[HEADER_GENERATION] = str(binding.generation)
    return headers


def checkpoint_thread_id(tenant_id: str | None, thread_id: str) -> str:
    """LangGraph checkpointer key for this assignment's logical thread.

    Empty / "default" tenants keep the bare thread_id so pre-tenancy and
    single-tenant direct-mode checkpoint rows stay reachable. Any other
    tenant gets "{tenant_id}:{thread_id}" so two tenants cannot share a
    checkpoint row when they reuse the same client-chosen thread id.
    """
    tid = (tenant_id or "").strip() or _DEFAULT
    if tid == _DEFAULT:
        return thread_id
    return f"{tid}:{thread_id}"
